portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,124 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+public class HighestPriorityFirstScheduler implements TaskScheduler {
+  private static final Comparator<ScheduledTask> sortByScheduledStartTimeAscending =
+      Comparator.comparing(ScheduledTask::getStartTime);
+
+  /**
+   * This method schedules tasks from highest to lowest priority and shortest to longest duration.
+   */
+  public Collection<ScheduledTask> schedule(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+
+    TaskQueue taskQueue = new TaskQueue(tasksList, getSchedulingAlgorithmType());
+
+    CalendarEventsGroup calendarEventsGroup =
+        new CalendarEventsGroup(eventsList, workHoursStartTime, workHoursEndTime);
+
+    List<TimeRange> availableTimes = calendarEventsGroup.getFreeTimeRanges();
```

Comment:
Maybe add a comment that says that we should work with currentAvailableTimes and not use availableTimes.

Reply:
Done.